dlc-manager/src/chain_monitor.rs

```diff
@@ -108,41 +107,91 @@ impl ChainMonitor {
         }
     }
 
-    pub(crate) fn process_block(
-        &self,
-        block: &Block,
-        height: u64,
-    ) -> Vec<(Transaction, ChannelInfo)> {
-        let mut res = Vec::new();
-
+    /// Check if any watched transactions are part of the block, confirming them if so.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the new block's height is not exactly one more than the last processed height.
+    pub(crate) fn process_block(&mut self, block: &Block, height: u64) {
         assert_eq!(self.last_height + 1, height);
 
-        for tx in &block.txdata {
-            let channel_info = self.watched_tx.get(&tx.txid()).or_else(|| {
-                for txid in tx.input.iter().map(|x| &x.previous_output.txid) {
```

Comment:
Didn't the `or_else` part disappear in your version?

Reply:
You are right! And it wasn't on purpose.

But looking at it now, what is the purpose of it? It seems like for every transaction in the block we:

1. Look it up in our hashmap.
2. If not in the hashmap, we look up every input to the transaction in the hashmap too.

Is the second step to ensure that we don't miss transactions? Given the assertion above, I don't think we can miss blocks without triggering the panic. Am I missing something?

Reply:
To be honest it might not be the prettiest, but the point is to detect both transactions that are registered, and transactions that spend those. It's required by the logic of the channel manager, to detect when a buffer transaction is spent, as it would be a pain to register all CETs. Feel free to refactor if you think of something better :)

Reply:
Oh, interesting! I hadn't considered that. It's obviously a bit surprising so I would just try to document it clearly for now.

In terms of a possible rework, I guess what we want is the ability to watch transactions _and_ transaction outputs. `rust-lightning` offers the same two options with the `chain::Filter` trait.

---

Btw, I thought about reusing `rust-lightning` for this stuff but I think it's not really possible, because those traits are built to be implemented by consumers but used by LDK, which makes sense. It would be pretty weird to plug into that, but we could copy the structure so that `ChainMonitor` can be replaced with a trait.

Reply:
> Oh, interesting! I hadn't considered that. It's obviously a bit surprising so I would just try to document it clearly for now.

Yeah definitely not super clear, sorry :)

> In terms of a possible rework, I guess what we want is the ability to watch transactions and transaction outputs. rust-lightning offers the same two options with the chain::Filter trait.

Yeah I guess that's it!

> but we could copy the structure so that ChainMonitor can be replaced with a trait.

Yeah that's definitely what I'd like to do!